# 3. Funções

Uma *função* é uma sequência de declarações que executam operações de computação. Quando define uma função, você especifica um *nome* e uma sequência de instruções. Depois, você pode *chamar*, ou *utilizar a função*, pelo nome que você especificou.

## Chamada de Funções

Nós já vimos um exemplo de chamada de função:

In [1]:
println("Olá, Mundo!")

Olá, Mundo!


O nome dessa função é `println`. A expressão entre parênteses é chamada de *argumento* da função.
É comum dizer que uma função *recebe* um argumento e *devolve* um resultado. O resultado é chamado de *valor de retorno*.

Julia possui funções que convertem valores de um tipo para outro. A função `parse`, por exemplo, recebe uma string e a converte em um número, se possível, ou exibe uma mensagem de erro caso contrário.

In [2]:
parse(Int64, "32")

32

In [3]:
parse(Float64, "3.14159")

3.14159

In [4]:
parse(Int64, "Olá")

ArgumentError: ArgumentError: invalid base 10 digit 'O' in "Olá"

No último caso não foi possível fazer a conversão, pois não é possível converter letras para inteiros. Nos dois primeiros casos, fizemos a conversão de uma string com números para inteiros e floats.

A função `trunc` pode converter números de ponto flutuante para números inteiros, mas ao invés de arredondar o valor, ela remove sua parte decimal:

In [5]:
trunc(Int64, 3.99999)

3

In [6]:
trunc(Int64, -2.3)

-2

A função `float` converte números inteiros para números de ponto flutuante:

In [7]:
float(32)

32.0

Por fim, a função `string` converte seu argumento para string:

In [8]:
string(32)

"32"

In [9]:
string(3.14159)

"3.14159"

## Funções Matemáticas

A maioria das funções matemáticas conhecidas está disponível para utilização direta em Julia:

O exemplo abaixo usa a função `log10` para calcular uma razão entre as intensidades de sinal e ruído em decibéis. A função `log`, que calcula logaritmos naturais, também está disponível.

In [5]:
intensidade_sinal = 15.848931924611142
intensidade_ruido = 0.001

razão = intensidade_sinal / intensidade_ruido
decibéis = 10 * log10(razão)

42.0

O próximo exemplo calcula o seno de `radianos` . O nome da variável é uma dica, pois `sin` e as outras funções trigonométricas, como `cos` e `tan`, recebem argumentos em radianos. Para converter *graus* para *radianos* , divida por 180 e multiplique por $\pi$, como mostrado nos exemplos seguintes:

In [6]:
radianos = 0.7
altura = sin(radianos)

0.644217687237691

In [9]:
graus = 45

45

In [3]:
radianos = graus / 180 * π

0.7853981633974483

In [7]:
sin(radianos)

0.644217687237691

> *Aviso*: Note que a constante $\pi$ pode ser usada sem definição prévia através dos símbolos $\pi$, obtido digitando `\pi` e apertando a tecla `TAB`, ou simplesmente digitando `pi`.

o valor da variável $\pi$ é uma aproximação em ponto flutuante para o número $\pi$ com precisão de aproximadamente 16 dígitos. Você pode checar o resultado anterior comparando-o com $\dfrac{\sqrt{2}}{2}$:

In [15]:
sqrt(2) / 2

0.7071067811865476

## Composição de Funções

Até agora, consideramos separadamente os elementos de um programa, isto é, *variáveis*, *expressões* e *declarações*, sem discutir como combiná-los.
Um dos recursos mais úteis em linguagens de programação é a possibilidade de combinar operações simples para construir operações mais complexas. Os argumentos passados a uma função podem ser expressões de qualquer tipo, incluindo operações aritméticas:

In [10]:
x = sin(graus / 360 * 2 * π)

0.7071067811865475

Os argumentos passados a uma função também podem ser *chamadas a outras funções*:

In [11]:
x = exp(log(x+1))

1.7071067811865475

Você pode usar expressões em quase todos os lugares onde pode usar valores, contanto que *o lado esquerdo de uma atribuição seja um nome de variável*. Qualquer outra expressão no lado esquerdo é considerada *erro de sintaxe*. Veremos exceções para essa regra mais tarde.

Considere o exemplo abaixo:

In [14]:
horas = 60 # 1 hora equivale a 60 minutos
minutos = horas * 60 # certo

3600

In [13]:
horas * 60 = minutos # errado!

ErrorException: syntax: "60" is not a valid function argument name

## Adicionando Novas Funções

Até agora, usamos apenas funções já disponíveis na linguagem Julia, mas também podemos adicionar nossas próprias funções. Uma *definição de função* define o nome de uma nova função e a sequência de instruções que será executada quando a função for utilizada.

Considere exemplo abaixo:

In [20]:
function exibemusica()
    println("Sou um lenhador, e tá tudo maneiro.")
    println("Durmo a noite inteira, e trabalho o dia inteiro.")
end

exibemusica (generic function with 1 method)

A palavra-chave `function` faz desse trecho de código uma definição de função. O nome da função nesse exemplo é `exibemusica`. As regras para nomes de função são as mesmas utilizadas para nomes de variáveis. Nomes de função podem conter quase todos os caracteres *Unicode*, mas *o primeiro caractere não pode ser um número*. Veja o capítulo Caracteres. Você não pode usar uma palavra-chave como nome de uma função, e você deve evitar dar nomes idênticos a variáveis e funções.

Parênteses vazios depois de um nome indicam que a função não recebe argumentos.
A primeira linha de uma definição de função é chamada de *cabeçalho*, e o resto da função é chamado de *corpo*.

O corpo de uma função é marcado pela palavra-chave `end`, e pode conter qualquer quantidade de instruções. O corpo de uma função deve ser indentado para melhorar sua legibilidade.

Deve-se utilizar "aspas duplas", geralmente localizadas perto da tecla de *número 1* no teclado. «Aspas angulares» não são permitidas em Julia.

Se você definir uma função no modo interativo, o REPL a indentará automaticamente. Se você esquecer de terminar, ou *fechar*, uma definição de função com a palavra-chave `end`, verá um erro:

In [15]:
function exibemusica()
    println("Sou um lenhador, e tá tudo maneiro.")

LoadError: syntax: incomplete: "function" at none:1 requires end

Para encerrar uma função, você deve inserir a palavra-chave `end`. Completando a célula acima, temos:

In [19]:
function exibemusica()
    println("Sou um lenhador, e tá tudo maneiro.")
    println("Durmo a noite inteira, e trabalho o dia inteiro")
end

exibemusica (generic function with 1 method)

A sintaxe para chamar funções definidas pelo usuário é a mesma utilizada para chamar funções pré-definidas na linguagem:

In [20]:
exibemusica()

Sou um lenhador, e tá tudo maneiro.
Durmo a noite inteira, e trabalho o dia inteiro


Após definir uma função, você pode usá-la dentro de outras funções. Por exemplo, para repetir o primeiro refrão nós podemos escrever a função `repetemusica` :

In [22]:
function repetemusica()
    exibemusica()
    exibemusica()
end

repetemusica (generic function with 1 method)

E depois chamar `repetemusica`:

In [23]:
repetemusica()

Sou um lenhador, e tá tudo maneiro.
Durmo a noite inteira, e trabalho o dia inteiro
Sou um lenhador, e tá tudo maneiro.
Durmo a noite inteira, e trabalho o dia inteiro


No entanto, não é exatamente assim que a música toca.

## Definições e Utilizações

Combinando os fragmentos de código das seções anteriores, o programa inteiro fica da seguinte maneira:

In [24]:
function exibemusica()
    println("Sou um lenhador, e tá tudo maneiro.")
    println("Durmo a noite inteira, e trabalho o dia inteiro.")
end

function repetemusica()
    exibemusica()
    exibemusica()
end

repetemusica()

Sou um lenhador, e tá tudo maneiro.
Durmo a noite inteira, e trabalho o dia inteiro.
Sou um lenhador, e tá tudo maneiro.
Durmo a noite inteira, e trabalho o dia inteiro.


Esse programa contém duas definições de funções: `exibemusica` e `repetemusica` . As definições de funções são executadas da mesma forma que outras instruções, mas o efeito é criar funções. As instruções dentro das funções não são executadas enquanto a função não é chamada, e a definição de função não gera saídas.

Como você deve estar pensando, você precisa criar uma função antes de usá-la. Em outras palavras, uma definição de função tem que ser executada antes da própria função ser chamada.

### Exercício 3-1

a) Mova a última linha do seguinte programa para o início, então a chamada da função aparecerá antes das definições. Execute o programa e veja a mensagem de erro que aparecerá.

b) Agora, mova a chamada de função de volta para o final do programa e mova a definição de `exibemusica` para depois da definição de `repetemusica` . O que acontece quando voce roda esse programa?

## Fluxo de Execução

Para garantir que uma função é definida antes de sua primeira utilização, você tem que conhecer a ordem em que as instruções são executadas, o que é chamado de *fluxo de execução*.

As execuções sempre começam na primeira instrução de um programa. As instruções são executadas uma por vez, na ordem do topo ao final.

> *Dica*: Conhecido como *paralelismo*, é a forma conhecida para se executar mais de uma instrução por vez.

As definições de função não alteram o fluxo de execução de um programa, mas lembre-se que as instruções dentro de uma função não são executadas antes de sua chamada.

Uma chamada de função pode ser entendida como um desvio no fluxo de execução. Ao invés de ir para a próxima instrução, o fluxo é direcionado para o corpo da função chamada, executa suas instruções, e então volta ao ponto onde o fluxo havia sido interrompido.

Isso parace simples, antes que você se lembre de que uma função pode chamar outra. Assim, no meio de uma função, o programa poderá executar as instruções de outra. Então, enquanto executa uma função, o programa poderá executar ainda outra função!

Por sorte, Julia é boa para acompanhar esse fluxo, então cada vez que uma função é encerrada, o programa mostra onde ocorreu o encerramente dentro de sua chamada. Quando o programa chega ao fim, ele é encerrado.

Em resumo, quando você lê um programa, nem sempre a melhor opção é ler do topo ao fim. As vezes, faz mais sentido você acompanhar o fluxo de execução.

## Parâmetros e Argumentos

Algumas das funções que temos visto requerem *argumentos*. Por exemplo, quando você chama `sin` você passa um número como argumento. Algumas funções recebem mais que um argumento, como por exemplo: `parse` recebe dois argumentos, um do tipo número e outro do tipo string.

Dentro de uma função, os argumentos são atribuídos a variáveis chamadas *parâmetros*. Aqui temos um exemplo de definição para uma função que recebe um *argumento*:

In [17]:
function exibeduasvezes(bruce)
    println(bruce)
    println(bruce)
end

exibeduasvezes (generic function with 1 method)

Essa função atribui o *argumento* para um *parâmetro* chamado `bruce`. Quando uma função é chamada, ela exibe o valor do parâmetro (não importa o que seja) duas vezes.

Essa função utiliza qualquer valor que pode ser exibido.

In [18]:
exibeduasvezes("Spam")

Spam
Spam


In [28]:
exibeduasvezes(42)

42
42


In [29]:
exibeduasvezes(π)

π = 3.1415926535897...
π = 3.1415926535897...


As mesmas regras de composição que se aplicam à funções nativas também se aplicam à funções criadas pelo programador, então nós podemos utilizar qualquer tipo de expressão como argumento para `exibeduasvezes`:

In [30]:
exibeduasvezes("Spam "^4)

Spam Spam Spam Spam 
Spam Spam Spam Spam 


In [31]:
exibeduasvezes(cos(π))

-1.0
-1.0


Os argumentos são calculados antes da chamada de função, então o exemplo de expressões como `"Spam "^4` e `cos($\pi$)` são calculadas apenas uma vez.

Você também pode usar uma variável como argumento:

In [32]:
michael = "Eric, a metade de uma abelha."

"Eric, a metade de uma abelha."

In [33]:
exibeduasvezes(michael)

Eric, a metade de uma abelha.
Eric, a metade de uma abelha.


O nome da variável que nós passamos como argumento (`michael`) não tem relação com o nome do parâmetro (`bruce`). Então não importa como o valor é chamado anteriormente (em relação à chamada de função). Aqui na função `exibeduasvezes`, nós chamamos todas as variáveis de  `bruce`.

## Variáveis e Parâmetros São Locais

Quando você cria uma variável dentro de uma função, ela é *local*, o que significa que ela existe apenas dentro daquela função. Por exemplo:

In [19]:
function concatenaexibe(parte1, parte2)
    concat = parte1 * parte2
    exibeduasvezes(concat)
end

concatenaexibe (generic function with 1 method)

Essa função recebe dois argumentos, os concatena e exibe o resultado duas vezes. Aqui temos um exemplo de sua utilização:

In [12]:
linha1 = "Bing tiddle "

"Bing tiddle "

In [14]:
linha2 = "tiddle bang."

"tiddle bang."

In [20]:
concatenaexibe(linha1, linha2)

Bing tiddle tiddle bang.
Bing tiddle tiddle bang.


Quando `concatenaexibe` é encerrada, a variável `concat` é destruída. Se nós tentarmos exibí-la, receberemos o seguinte erro:

In [21]:
println(concat)

UndefVarError: UndefVarError: concat not defined

Os parâmetros também são locais. Por exemplo, fora de `exibeduasvezes`, não existe nada chamado `bruce`.

## Diagramas de Pilha

Para acompanhar e saber onde cada variável pode ser usada, as vezes é útil desenhar um *diagrama de pilha*. Como um diagrama de estados, um diagrama de pilha mostra os valores de cada variável e as funções as quais cada variável pertence.

Cada função é representada como um *quadro*. Um quadro é uma caixa com o nome da função, ao lado, e seus parâmetros e variáveis dentro. Veja alguns exemplos:

![diagrama de pilha](../docs/img/cap3-fig2-stack.png)
*Figura 2. Diagrama de pilha*

Os quadros são organizados em uma pilha que indica o que cada função chama. Nesse exemplo, `exibeduasvezes` foi chamada por `concatenaexibe`, e `concatenaexibe` foi chamada pela `Main`, que é um nome especial para o quadro que fica no topo. Quando você cria uma variável fora de qualquer função, ela pertence a função `Main`.

Cada parâmetro se refere ao mesmo valor que seu argumento correspondente. Então, `parte1` tem o mesmo valor que `linha1`, `parte2` tem o mesmo valor que `linha2`, e `bruce` tem o mesmo valor que `concat` .

Se ocorrer algum erro durante uma chamada de função, Julia exibe o nome da função onde houve o erro, o nome da função que a chamou, e assim por diante, até que chegue à função `Main` .

Por exemplo, se você tentar acessar a variável `concat` de dentro da função `exibeduasvezes`, você verá a seguinte mensagem de erro:

ERROR: UndefVarError: concat not defined        
Stacktrace:                                     
 [1] exibeduasvezes at ./REPL[1]:2 [inlined]         
 [2] concatenaexibe(::String, ::String) at ./REPL[2]:3

Essa lista de funções é chamada de *stacktrace*. Ela te mostra em qual arquivo do programa o erro ocorreu, assim como a linha e quais funções estavam sendo executadas no momento do erro.

A ordem das funções na stacktrace é inversa à oderm dos quadros no diagrama de pilhas. A função que estava sendo executada no momento do erro é a que está no topo.

## Funções com Resultado e Funções Nulas

Algumas das funções que nós temos usado, tais como as funções matemáticas, retornam resultados (valores). Por falta de um nome melhor, eu as chamo de *Funções Frutíferas*. Outras funções como `exibeduasvezes` que não devolvem resultados, elas são chamadas de *Funções nulas* .

Quanto você chama uma função frutífera, você quase sempre quer utilizar o resultado para alguma coisa. Por exemplo, você pode salvar esse valor em uma variável e usá-lo como parte de uma expressão matemática:

## Funções Frutíferas e Funções Nulas

Algumas das funções que nós temos usado, tais como as funções matemáticas, retornam resultados (valores). Por falta de um nome melhor, eu as chamo de *Funções Frutíferas*. Outras funções como `exibeduasvezes` que não devolvem resultados, elas são chamadas de *Funções nulas* .

Quanto você chama uma função frutífera, você quase sempre quer utilizar o resultado para alguma coisa. Por exemplo, você pode salvar esse valor em uma variável e usá-lo como parte de uma expressão matemática:

In [39]:
x = cos(radians)
golden = (sqrt(5) + 1) / 2

1.618033988749895

Quando você chama uma função no modo interativo, Julia mostra o resultado:

In [40]:
sqrt(5)

2.23606797749979

Esse script calcula a raiz quadrada de 5, mas, como nós não salvamos ou imprimimos o resultado, esse script não é muito útil.

Funções nulas podem imprimir algum resultado na tela ou ter algum outro efeito, mas elas não possuem valor de retorno. Se você salva seu resultado em uma variável, você verá um valor especial chamado `nothing`

Mas em um script, se você chama uma função frutífera sem salvar seu valor de retorno, o valor será perdido para sempre!

`sqrt(5)`

Output:

`2.23606797749979`

Mas em um script, se você chama uma função frutífera sem salvar seu valor de retorno, o valor será perdido para sempre!

`sqrt(5)`

Output:

`2.23606797749979`

In [41]:
result = exibeduasvezes("Bing")

Bing
Bing


In [42]:
show(result)

nothing

Para imprimir o valor `nothing`, você tem que usar a função `show` que funciona como a função `print` mas pode exibir o valor `nothing`.

O valor `nothing` não é o mesmo que a string `"nothing"`. Esse é um valor especial, com seu próprio tipo:

In [43]:
typeof(nothing)

Nothing

In [44]:
typeof("nothing")

String

As funções que nós temos escrito até agora são todas nulas. Nós começaremos a escrever funções frutíferas em alguns capítulos.

## Por que Funções?

Pode não estar claro o porque do trabalho de dividir um programa em funções. Aqui temos alguns motivos:

- Criar novas funções te permite nomear um grupo de instruções, o que torna seu programa mais fácil de ser lido e depurado.
- As funções podem deixar seu programa menor, eliminando trechos repetitivos de código. Mais tarde, se você fizer alguma modificação no código, você só precisará fazer tal alteração em apenas um local.
- Dividir um programa longo em funções te permite depurá-lo um pedaço por vez, e depois montá-los novamente.
- Funções bem estruturadas geralmente são utilizadas por vários programas. Uma vez que você a escreveu e depurou, você pode reutilizá-la.
- Em Julia, funções podem otimizar muito a performance de um programa.

## Depurando

Uma das habilidades mais importantes que você irá adiquirir é *depurar*. Embora possa ser frustante, depurar é uma das partes mais intelectualmente ricas, desafiadoras, e interessantes da programação.

De certa forma, depurar é como o trabalho de um detetive. Você possui pistas e tem que inferir quais processos e eventos estão levando aos resultados que você recebe.

Depurar também é como uma ciência experimental. Uma vez que você tem alguma ideia do porque as coisas estão dando errado, você modifica o seu programa e tenta novamente. Se suas hipóteses estavam corretas, você pode prever o resultado de suas modificações, e então chegar um passo mais perto de ter seu programa funcionando. Se suas hipóteses estavam erradas, você tem que formular uma nova. 

Como Sherlock Holmes salientou: "*Quando você eliminar o impossível, o que restar, mesmo que improvável, deve ser a verdade.*" - A. Conan Doyle, The Sign of Four.

Para algumas pessoas, programação e depuração é a mesma coisa. Ou seja, programar é o processo gradual de depurar um programa até que ele faça o que você quer. A ideia é que você comece com um programa funcional e faça pequenas modificações, depurando-as conforme avançar.

Por exemplo, o Linux é um sistema operacional que contém milhões de linhas de código, mas ele foi iniciado como um programa simples que Linus Torvalds utilizava para explorar o microprocessador Intel 80386. De acordo com Larry Greenfield, "*Um dos primeiros projetos do Linus era um programa que alternava entre exibir "AAAA" e "BBBB". Mais tarde, isso evoluiu para o Linux.*" (*The Linux User's Guide* Beta Version 1).

## Glossário

**função**

Uma sequência (nomeada) de instruções que calculam alguma operação útil. As funções podem ou não receber argumentos e podem ou não retornar resultados.

**definição de função**

Uma instrução que cria novas funções, especificando seu nome, parâmentros, e as intruções que ela contém.

**objeto de função**

O valor criado por uma definição de função. O nome de uma função é uma variável que se refere a um objeto de função.

**cabeçalho**

A primeira linha de uma definição de função.

**corpo**

A sequência de instruções dentro de uma definição de função.

**parâmetro**

Um nome usado dentro de uma função que se refere aos valores passados como argumento.

**chamada de função**

A instrução que executa uma função. Consiste no nome da função seguido de uma lista de parâmetros entre parênteses.

**argumento**

Um valor passado para uma função no momento de sua chamada. Esse valor é atribuído a um parâmetro correspondente dentro da função.

**variável local**

Uma variável definida dentro de uma função. Uma variável local pode ser usada apenas dentro de sua função.

**função frutífera**

Uma função que retorna (ou devolve) um resultado.

**função nula**

Uma função que sempre retorna `nothing`

`nothing`

Um valor especial retornado por funções nulas.

**composição**

O uso de uma expressão como parte de uma outra maior, ou uma instrução compondo uma instrução maior.

**fluxo de execução**

A ordem de intruções executadas.

**diagrama de pilha**

Uma representação gráfica de uma pilha de funções, suas variáveis, e os valores aos quais se referem.

**quadro**

Uma caixa em um diagrama de pilha representando uma chamada de função. Ela contém as variáveis locais e os parâmetros de uma função.

**stacktrace**

Uma lista de funções que estão sendo executadas, exibidas quando um erro ocorre.

## Exercícios

> *Dica*: 
> Esses exercícios devem ser feitos utilizando apenas as instruções e os outros recursos que temos aprendido até agora.

### Exercício 3-2

Escreva uma função chamada `jutificadireita` que recebe uma string chamada `s` como parâmetro e exibe a string com espaços suficientes para que a última letra da string esteja na coluna 70 da tela. Por exemplo:

`justificadireita("monty")`
`monty`

> *Dica*: Use concatenação e repetição de strings. Além disso, Julia oferece uma função nativa chamada `lenght` que retorna o tamanho de uma string, então o valor de `lenght("monty")` é 5.

### Exercício 3-3

Um objeto de função é um valor que pode ser atribuído à uma variável ou passado como um argumento. Por exemplo, `facaduasvezes` é uma função que recebe um objeto de função como um argumento e o chama duas vezes:

In [1]:
function facaduasvezes(f)
    f()
    f()
end

facaduasvezes (generic function with 1 method)

Aqui temos um exemplo em que `facaduasvezes` chama a função `imprimespam` duas vezes:

In [2]:
function imprimespam()
    println("spam")
end

facaduasvezes(imprimespam)

spam
spam


1. Use esse exemplo em um script e teste-o.
2. Modifique a função `facaduasvezes` para que ela receba dois argumentos, um objeto de função, um valor, e chame essa função duas vezes, passando o valor como um argumento.
3. Copie a definição da função `exibeduasvezes`, que vimos mais cedo nesse capítulo, para seu script.
4. Use a versão modificada de `facaduasvezes` para chamar a função `exibeduasvezes` duas vezes, passando `"spam"` como um argumento.
5. Crie uma nova função chamada `facaquatrovezes` que recebe um objeto de função, um valor e chame a função quatro vezes, passando o valor como um parâmetro. Devem haver apenas duas intruções no corpo dessa função, não quatro.

### Exercício 3-4

1. Escreva a função `imprimegrade` que desenhe uma grade como o exemplo a seguir:

`imprimegrade()`
![grid](../docs/img/cap3-exec3.png)

2. Escreva uma função que desenhe uma grade similar om quatro linhas e quatro colunhas.


Créditos: Esses exercícios são baseados no livro Oualline, Practical C Programming, Third Edition, O’Reilly Media, 1997.

> *Dica*: Para imprimir mais de um valor em uma linha, você pode imprimir uma vírgula como separador de uma sequência de valores: `println("+", "-")`. A função `print` não avança para a próxima linha, por exemplo: `print("+ ") println("-")` exibirá a saída `"+ -"`, na mesma linha. A saída da próxima instrução de print começará na próxima linha, pois a última instrução foi `println` ao invés de `print`.

